alf/algorithms/functional_particle_vi_algorithm_test.py

```diff
@@ -0,0 +1,169 @@
+                a dimension while the rows contains observations.
+
+        Returns:
+            The covariance matrix
+        """
+        x = data.detach().clone()
+        if x.dim() > 2:
+            raise ValueError('data has more than 2 dimensions')
+        if x.dim() < 2:
+            x = x.view(1, -1)
+        if not rowvar and x.size(0) != 1:
+            x = x.t()
+        fact = 1.0 / (x.size(1) - 1)
+        x -= torch.mean(x, dim=1, keepdim=True)
+        return fact * x.matmul(x.t()).squeeze()
+
+    def assertArrayGreater(self, x, y, eps):
+        self.assertEqual(x.shape, y.shape)
+        self.assertGreater(float(torch.min(x - y)), eps)
+
+    @parameterized.parameters(('gfsf'), ('svgd'), ('gfsf', True),
+                              ('svgd', True))
+    def test_functional_par_vi_algorithm(self,
+                                         par_vi='svgd',
+                                         function_vi=False,
+                                         num_particles=256,
+                                         train_batch_size=10):
+        """
+        The hypernetwork is trained to generate the parameter vector for a linear
+        regressor. The target linear regressor is :math:`y = X\beta + e`, where 
+        :math:`e\sim N(0, I)` is random noise, :math:`X` is the input data matrix, 
+        and :math:`y` is target ouputs. The posterior of :math:`\beta` has a 
+        closed-form :math:`p(\beta|X,y)\sim N((X^TX)^{-1}X^Ty, X^TX)`.
+        For a linear generator with weight W and bias b, and takes standard Gaussian 
+        noise as input, the output follows a Gaussian :math:`N(b, WW^T)`, which should 
+        match the posterior :math:`p(\beta|X,y)` for both svgd and gfsf.
+        
+        """
+        input_size = 3
+        input_spec = TensorSpec((input_size, ), torch.float32)
+        output_dim = 1
+        batch_size = 100
+        inputs = input_spec.randn(outer_dims=(batch_size, ))
+        beta = torch.rand(input_size, output_dim) + 5.
+        print("beta: {}".format(beta))
+        noise = torch.randn(batch_size, output_dim)
+        targets = inputs @ beta + noise
+        true_cov = torch.inverse(
+            inputs.t() @ inputs)  # + torch.eye(input_size))
+        true_mean = true_cov @ inputs.t() @ targets
+        noise_dim = 3
+        algorithm = FuncParVIAlgorithm(
+            input_tensor_spec=input_spec,
+            last_layer_param=(output_dim, False),
+            last_activation=math_ops.identity,
+            num_particles=num_particles,
+            loss_type='regression',
+            par_vi=par_vi,
+            function_vi=function_vi,
+            function_bs=train_batch_size,
+            optimizer=alf.optimizers.Adam(lr=1e-2))
+        print("ground truth mean: {}".format(true_mean))
+        print("ground truth cov: {}".format(true_cov))
+        print("ground truth cov norm: {}".format(true_cov.norm()))
+
+        def _train(train_batch=None, entropy_regularization=None):
```

Comment:
train_batch and entropy_regularization is not used by the caller, they can be removed.

Reply:
Updated.